tls/s2n_connection_serialize.c

```diff
@@ -0,0 +1,284 @@
+    /* No other version is supported currently */
+    RESULT_ENSURE_EQ(serialized_version, S2N_SERIALIZED_CONN_V1);
+
+    uint8_t protocol_version[S2N_TLS_PROTOCOL_VERSION_LEN] = { 0 };
+    RESULT_GUARD_POSIX(s2n_stuffer_read_bytes(&input, protocol_version, S2N_TLS_PROTOCOL_VERSION_LEN));
+    temp->protocol_version = (protocol_version[0] * 10) + protocol_version[1];
+
+    uint8_t cipher_suite[S2N_TLS_CIPHER_SUITE_LEN] = { 0 };
+    RESULT_GUARD_POSIX(s2n_stuffer_read_bytes(&input, cipher_suite, S2N_TLS_CIPHER_SUITE_LEN));
+    RESULT_GUARD(s2n_cipher_suite_from_iana(cipher_suite, S2N_TLS_CIPHER_SUITE_LEN, &temp->cipher_suite));
+
+    RESULT_GUARD_POSIX(s2n_stuffer_read_bytes(&input, temp->client_sequence_number, S2N_TLS_SEQUENCE_NUM_LEN));
+    RESULT_GUARD_POSIX(s2n_stuffer_read_bytes(&input, temp->server_sequence_number, S2N_TLS_SEQUENCE_NUM_LEN));
+
+    RESULT_GUARD_POSIX(s2n_stuffer_read_uint16(&input, &temp->max_fragment_len));
+
+    if (temp->protocol_version >= S2N_TLS13) {
+        RESULT_GUARD(s2n_connection_deserialize_tls13_secrets(&input, temp));
+    } else {
+        RESULT_GUARD(s2n_connection_deserialize_secrets(&input, temp));
+    }
+
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_restore_tls13_secrets(struct s2n_connection *conn, struct s2n_connection_deserialize *temp)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(temp);
+
+    RESULT_CHECKED_MEMCPY(conn->secrets.version.tls13.client_app_secret,
+            temp->client_application_secret, temp->secret_size);
+    RESULT_CHECKED_MEMCPY(conn->secrets.version.tls13.server_app_secret,
+            temp->server_application_secret, temp->secret_size);
+    RESULT_CHECKED_MEMCPY(conn->secrets.version.tls13.resumption_master_secret,
+            temp->resumption_master_secret, temp->secret_size);
+
+    RESULT_GUARD(s2n_tls13_key_schedule_set_key(conn, S2N_MASTER_SECRET, S2N_SERVER));
+    RESULT_GUARD(s2n_tls13_key_schedule_set_key(conn, S2N_MASTER_SECRET, S2N_CLIENT));
+
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_restore_secrets(struct s2n_connection *conn, struct s2n_connection_deserialize *temp)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(temp);
+
+    RESULT_CHECKED_MEMCPY(conn->secrets.version.tls12.master_secret, temp->master_secret,
+            S2N_TLS_SECRET_LEN);
+    RESULT_CHECKED_MEMCPY(conn->handshake_params.client_random, temp->client_random,
+            S2N_TLS_RANDOM_DATA_LEN);
+    RESULT_CHECKED_MEMCPY(conn->handshake_params.server_random, temp->server_random,
+            S2N_TLS_RANDOM_DATA_LEN);
+    RESULT_GUARD_POSIX(s2n_prf_key_expansion(conn));
+
+    return S2N_RESULT_OK;
+}
+
+int s2n_connection_deserialize(struct s2n_connection *conn, uint8_t *buffer, uint32_t buffer_length)
+{
+    POSIX_ENSURE_REF(conn);
+    POSIX_ENSURE_REF(conn->secure);
+
+    /* Read parsed values into a temporary struct so that the connection is unaltered if parsing fails */
+    struct s2n_connection_deserialize temp = { 0 };
```

Comment:
"temp" isn't a very descriptive name :) Especially when it gets passed down to other methods that don't know / care about its lifetime.